contracts/BSCValidatorSet.sol

```diff
@@ -341,63 +367,49 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
 
   /*********************** For slash **************************/
   function misdemeanor(address validator)external onlySlash override{
-    uint256 index = currentValidatorSetMap[validator];
-    if (index <= 0) {
-      return;
-    }
-    // the actually index
-    index = index - 1;
-    uint256 income = currentValidatorSet[index].incoming;
-    currentValidatorSet[index].incoming = 0;
-    uint256 rest = currentValidatorSet.length - 1;
-    emit validatorMisdemeanor(validator,income);
-    if (rest==0) {
-      // should not happen, but still protect
-      return;
+    uint256 validatorIndex = _misdemeanor(validator);
+    if (canEnterMaintenance(validatorIndex)) {
+      _enterMaintenance(validator, validatorIndex);
     }
-    uint256 averageDistribute = income/rest;
-    if (averageDistribute!=0) {
-      for (uint i=0;i<index;i++) {
-        currentValidatorSet[i].incoming = currentValidatorSet[i].incoming + averageDistribute;
-      }
-      uint n = currentValidatorSet.length;
-      for (uint i=index+1;i<n;i++) {
-        currentValidatorSet[i].incoming = currentValidatorSet[i].incoming + averageDistribute;
-      }
-    }
-    // averageDistribute*rest may less than income, but it is ok, the dust income will go to system reward eventually.
   }
 
-  function felony(address validator)external onlySlash override{
-    uint256 index = currentValidatorSetMap[validator];
-    if (index <= 0) {
-      return;
+  function felony(address validator)external onlySlash override {
+    _felony(validator);
+  }
+
+  /*********************** For Temporary Maintenance **************************/
+  function canEnterMaintenance(uint256 index) public view returns (bool) {
+    Validator memory validatorInfo = currentValidatorSet[index];
+    address validator = validatorInfo.consensusAddress;
```

Comment:
We can pass `validator` as a parameter of `canEnterMaintenance`

Reply:
For example, if index is `100`, it will return true.

Reply:
https://github.com/binance-chain/bsc-genesis-contract/pull/120#discussion_r776134581

Reply:
fixed

Reply:
> Please pass `validator` as a parameter rather than `index`. For current implementation, it will return invalid result if the index is quite large.

 Passing `validator` as a parameter  will cause a lot to change, and the `pragma experimental ABIEncoderV2;
` needs to be imported.
So I've now added the index check to fix this problem